In [24]:
import pandas as pd

# Load the dataset
file_path = '/content/KL_encoded_data.csv'
df = pd.read_csv(file_path)
data=df
# Display the first few rows of the dataframe
data.head()


Bathrooms  Car Parks      Size  Type  No_of_Bedrooms  No_of_servant_rooms  \
0          3          2  1335.035     0               2                    1   
1          4          2  1875.075     0               3                    0   
2          3          2  1513.013     0               4                    1   
3          5          4  7200.000     1               4                    2   
4          4          4  3600.000     1               5                    0   

   studio_apartment  Location_Ampang  Location_Ampang Hilir  \
0                 0              0.0                    0.0   
1                 0              0.0                    0.0   
2                 0              0.0                    0.0   
3                 0              0.0                    0.0   
4                 0              0.0                    0.0   

   Location_Bandar Damai Perdana  ...  Property Type Supergroup_Townhouse  \
0                            0.0  ...                                 0.0   
1                            0.0  ...                                 0.0   
2                            0.0  ...                                 0.0   
3                            0.0  ...                                 0.0   
4                            0.0  ...                                 0.0   

   type_Bronze  type_Deluxe  type_Gold  type_Silver  size_type_Extra Large  \
0          0.0          0.0        1.0          0.0                    0.0   
1          0.0          0.0        0.0          1.0                    0.0   
2          0.0          0.0        0.0          1.0                    0.0   
3          0.0          1.0        0.0          0.0                    1.0   
4          0.0          0.0        1.0          0.0                    0.0   

   size_type_Large  size_type_Medium  size_type_Small    Price  
0              0.0               1.0              0.0  1250000  
1              1.0               0.0              0.0  1030000  
2              1.0               0.0              0.0   900000  
3              0.0               0.0              0.0  5350000  
4              1.0               0.0              0.0  2600000  

[5 rows x 87 columns]

In [25]:
X = df.drop('Price', axis=1)
y = df['Price']
X.head()
y.head()


0    1250000
1    1030000
2     900000
3    5350000
4    2600000
Name: Price, dtype: int64

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Lasso, Ridge, ElasticNet
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

# Assuming X and y are your features and target variable, respectively
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Linear Regression
linear_reg = LinearRegression()
linear_reg.fit(X_train, y_train)

# Lasso Regression
lasso_reg = Lasso()
lasso_reg.fit(X_train, y_train)

# Ridge Regression
ridge_reg = Ridge()
ridge_reg.fit(X_train, y_train)

# Elastic Net Regression
elastic_net_reg = ElasticNet()
elastic_net_reg.fit(X_train, y_train)

# Function to print evaluation metrics
def print_metrics(model, X_train, X_test, y_train, y_test, model_name):
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)
    print(f"{model_name} Metrics:")
    print("Training Set:")
    print("Mean Squared Error (MSE):", mean_squared_error(y_train, y_train_pred))
    print("Mean Absolute Error (MAE):", mean_absolute_error(y_train, y_train_pred))
    print("R^2 Score:", r2_score(y_train, y_train_pred))
    print("Testing Set:")
    print("Mean Squared Error (MSE):", mean_squared_error(y_test, y_test_pred))
    print("Mean Absolute Error (MAE):", mean_absolute_error(y_test, y_test_pred))
    print("R^2 Score:", r2_score(y_test, y_test_pred))
    print()

# Print evaluation metrics for each model
print_metrics(linear_reg, X_train, X_test, y_train, y_test, "Linear Regression")
print_metrics(lasso_reg, X_train, X_test, y_train, y_test, "Lasso Regression")
print_metrics(ridge_reg, X_train, X_test, y_train, y_test, "Ridge Regression")
print_metrics(elastic_net_reg, X_train, X_test, y_train, y_test, "Elastic Net Regression")


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_coordinate_descent.py:631: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 7.149e+15, tolerance: 1.261e+13
  model = cd_fast.enet_coordinate_descent(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_ridge.py:216: LinAlgWarning: Ill-conditioned matrix (rcond=1.85055e-17): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T


Linear Regression Metrics:
Training Set:
Mean Squared Error (MSE): 1924783782898.505
Mean Absolute Error (MAE): 467310.493523811
R^2 Score: 0.6459922309617542
Testing Set:
Mean Squared Error (MSE): 1493170393094.9104
Mean Absolute Error (MAE): 485989.3604012088
R^2 Score: 0.6896367481201522

Lasso Regression Metrics:
Training Set:
Mean Squared Error (MSE): 1924783809524.1062
Mean Absolute Error (MAE): 467294.58699710126
R^2 Score: 0.6459922260647524
Testing Set:
Mean Squared Error (MSE): 1493114778107.6528
Mean Absolute Error (MAE): 485971.2417852425
R^2 Score: 0.6896483079852408

Ridge Regression Metrics:
Training Set:
Mean Squared Error (MSE): 1925363077422.6777
Mean Absolute Error (MAE): 467047.1934695316
R^2 Score: 0.6458856866506786
Testing Set:
Mean Squared Error (MSE): 1470108629185.553
Mean Absolute Error (MAE): 484781.8025214021
R^2 Score: 0.6944302559971454

Elastic Net Regression Metrics:
Training Set:
Mean Squared Error (MSE): 2818436473796.131
Mean Absolute Error (MAE): 69

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from math import sqrt

# Function to print evaluation metrics for train-test split
def print_train_test_metrics(model, X_train, X_test, y_train, y_test, model_name):
    model.fit(X_train, y_train)
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)

    print(f"==================== 1-Fold Train-Test Split ====================")
    print(f"Model: {model_name}")
    print("Training Set:")
    print(f"R^2 score: {r2_score(y_train, y_train_pred)}")
    print(f"MSE: {mean_squared_error(y_train, y_train_pred)}")
    print(f"MAE: {mean_absolute_error(y_train, y_train_pred)}")
    print(f"RMSE: {sqrt(mean_squared_error(y_train, y_train_pred))}")
    print("Testing Set:")
    print(f"R^2 score: {r2_score(y_test, y_test_pred)}")
    print(f"MSE: {mean_squared_error(y_test, y_test_pred)}")
    print(f"MAE: {mean_absolute_error(y_test, y_test_pred)}")
    print(f"RMSE: {sqrt(mean_squared_error(y_test, y_test_pred))}")
    print()

# Print evaluation metrics for each model using train-test split
for model_name, pipeline in pipelines.items():
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
    print_train_test_metrics(pipeline, X_train, X_test, y_train, y_test, model_name)


==================== 1-Fold Train-Test Split ====================
Model: Linear Regression
Training Set:
R^2 score: 0.6230038501780222
MSE: 2049774436769.8403
MAE: 457675.1370270416
RMSE: 1431703.3340639535
Testing Set:
R^2 score: 0.6582942666404464
MSE: 1643960362938.8271
MAE: 479504.5254438218
RMSE: 1282170.1770587347

==================== 1-Fold Train-Test Split ====================
Model: Ridge Regression
Training Set:
R^2 score: 0.6230654230257476
MSE: 2049439657623.3523
MAE: 456713.80288066994
RMSE: 1431586.4129081948
Testing Set:
R^2 score: 0.657809476463848
MSE: 1646292708453.8828
MAE: 478340.3053043777
RMSE: 1283079.385094267

==================== 1-Fold Train-Test Split ====================
Model: Random Forest Regression
Training Set:
R^2 score: 0.9420785804126398
MSE: 314925882579.05194
MAE: 116436.20719331554
RMSE: 561182.5750850181
Testing Set:
R^2 score: 0.7551966545656525
MSE: 1177758981835.4656
MAE: 226620.01861623797
RMSE: 1085246.0466804132

==================== 1-Fo

In [ ]:
from sklearn.model_selection import KFold, cross_val_score
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from math import sqrt

# Function to print evaluation metrics for k-fold cross-validation
def print_cv_metrics(model, X, y, model_name, fold):
    kf = KFold(n_splits=fold, shuffle=True, random_state=42)
    r2_scores = cross_val_score(model, X, y, cv=kf, scoring='r2')
    mse_scores = cross_val_score(model, X, y, cv=kf, scoring='neg_mean_squared_error')
    mae_scores = cross_val_score(model, X, y, cv=kf, scoring='neg_mean_absolute_error')
    rmse_scores = [sqrt(abs(score)) for score in mse_scores]

    print(f"==================== {fold}-Fold Cross Validation ====================")
    print(f"Model: {model_name}")
    print(f"R^2 scores: {r2_scores}")
    print(f"Mean R^2: {r2_scores.mean()}")
    print(f"MSE scores: {mse_scores}")
    print(f"Mean MSE: {mse_scores.mean()}")
    print(f"MAE scores: {mae_scores}")
    print(f"Mean MAE: {mae_scores.mean()}")
    print(f"RMSE scores: {rmse_scores}")
    print(f"Mean RMSE: {sum(rmse_scores) / len(rmse_scores)}")
    print()

# Print evaluation metrics for each model using k-fold cross-validation
for fold in [3, 5]:
    for model_name, pipeline in pipelines.items():
        print_cv_metrics(pipeline, X, y, model_name, fold)


==================== 3-Fold Cross Validation ====================
Model: Linear Regression
R^2 scores: [0.54731079 0.66200395 0.68186736]
Mean R^2: 0.6303940348135121
MSE scores: [-2.90437780e+12 -1.78930371e+12 -1.34371093e+12]
Mean MSE: -2012464148937.3142
MAE scores: [-468757.35452406 -472901.26010435 -473354.05893944]
Mean MAE: -471670.89118928305
RMSE scores: [1704223.519269203, 1337648.5755147547, 1159185.4604009583]
Mean RMSE: 1400352.5183949722

==================== 3-Fold Cross Validation ====================
Model: Ridge Regression
R^2 scores: [0.54773113 0.66363817 0.68644853]
Mean R^2: 0.6326059450500453
MSE scores: [-2.90168098e+12 -1.78065238e+12 -1.32436123e+12]
Mean MSE: -2002231531449.1897
MAE scores: [-466824.22223101 -466833.5056226  -444566.04061418]
Mean MAE: -459407.9228225965
RMSE scores: [1703432.119023468, 1334410.8735997374, 1150808.9461963582]
Mean RMSE: 1396217.3129398546

==================== 3-Fold Cross Validation ====================
Model: Random Forest

In [26]:
from sklearn.tree import DecisionTreeRegressor

# Decision Tree Regression
decision_tree_reg = DecisionTreeRegressor(random_state=42)
decision_tree_reg.fit(X_train, y_train)

# Print evaluation metrics for decision tree regression
print_metrics(decision_tree_reg, X_train, X_test, y_train, y_test, "Decision Tree Regression")


Decision Tree Regression Metrics:
Training Set:
Mean Squared Error (MSE): 44139516832.99141
Mean Absolute Error (MAE): 26537.129819674778
R^2 Score: 0.99188182484739
Testing Set:
Mean Squared Error (MSE): 1346729551710.0767
Mean Absolute Error (MAE): 264960.61854562996
R^2 Score: 0.7200752405724526



In [31]:
import warnings
from sklearn.metrics import r2_score

# Suppress convergence warnings
warnings.filterwarnings("ignore")

# Define a function to print cross-validation scores
def print_cross_val_scores(model, X, y, num_folds, model_name):
    kf = KFold(n_splits=num_folds, shuffle=True, random_state=42)
    mse_scores = cross_val_score(model, X, y, cv=kf, scoring='neg_mean_squared_error')
    r2_scores = cross_val_score(model, X, y, cv=kf, scoring='r2')

    print(f"{model_name} Cross-Validation Scores (k={num_folds}):")
    print("Mean Squared Error (MSE): {:.2e}".format(-mse_scores.mean()))
    print("Standard Deviation (MSE): {:.2e}".format(mse_scores.std()))
    print("R-squared (R2) Score: {:.2f}".format(r2_scores.mean()))
    print()

# Define your models
models = {
    "Linear Regression": linear_reg,
    "Lasso Regression": lasso_reg,
    "Ridge Regression": ridge_reg,
    "Elastic Net Regression": elastic_net_reg,
    "Decision Tree Regression": decision_tree_reg
}

# Perform cross-validation for each model with different numbers of folds
for model_name, model in models.items():
    for num_folds in [3, 5]:
        print_cross_val_scores(model, X, y, num_folds, model_name)


Linear Regression Cross-Validation Scores (k=3):
Mean Squared Error (MSE): 1.90e+12
Standard Deviation (MSE): 7.14e+11
R-squared (R2) Score: 0.65

Linear Regression Cross-Validation Scores (k=5):
Mean Squared Error (MSE): 1.93e+12
Standard Deviation (MSE): 1.06e+12
R-squared (R2) Score: 0.66

Lasso Regression Cross-Validation Scores (k=3):
Mean Squared Error (MSE): 1.90e+12
Standard Deviation (MSE): 7.14e+11
R-squared (R2) Score: 0.65

Lasso Regression Cross-Validation Scores (k=5):
Mean Squared Error (MSE): 1.93e+12
Standard Deviation (MSE): 1.06e+12
R-squared (R2) Score: 0.66

Ridge Regression Cross-Validation Scores (k=3):
Mean Squared Error (MSE): 1.90e+12
Standard Deviation (MSE): 7.18e+11
R-squared (R2) Score: 0.65

Ridge Regression Cross-Validation Scores (k=5):
Mean Squared Error (MSE): 1.93e+12
Standard Deviation (MSE): 1.07e+12
R-squared (R2) Score: 0.66

Elastic Net Regression Cross-Validation Scores (k=3):
Mean Squared Error (MSE): 2.69e+12
Standard Deviation (MSE): 8.65e+1

the performance of different regression models using various evaluation metrics such as Mean Squared Error (MSE), R-squared (R2) score, and Standard Deviation of MSE. Let's break down the analysis:

Decision Tree Regression Metrics:

The Decision Tree Regression model achieves very low MSE on the training set (44139516832.99) but significantly higher MSE on the testing set (1346729551710.08). This indicates that the model is overfitting the training data, as it performs well on data it has seen before but generalizes poorly to unseen data.
The R-squared (R2) score on the testing set (0.720) indicates that the model explains about 72.01% of the variance in the target variable, which is relatively good but could be improved.
The significant difference between training and testing MSE suggests that the model may benefit from regularization techniques or other methods to reduce overfitting.
Cross-Validation Scores:

Linear Regression, Lasso Regression, and Ridge Regression models exhibit similar performance across different numbers of folds (k=3 and k=5). They have relatively high MSE values, indicating moderate predictive accuracy.
The R2 scores for these models range from 0.65 to 0.66, indicating that they explain around 65-66% of the variance in the target variable. This suggests moderate predictive power but leaves room for improvement.
Elastic Net Regression performs slightly worse compared to other models, with lower R2 scores around 0.51 to 0.52. This indicates that the model explains less variance in the target variable compared to other models.
Comparison between Models:

Decision Tree Regression has the lowest MSE on the training set but performs poorly on the testing set, suggesting overfitting.
Linear Regression, Lasso Regression, and Ridge Regression models show consistent performance with moderate predictive accuracy and similar R2 scores.
Elastic Net Regression performs slightly worse than the other models, indicating less predictive power.